### DISCUSSION TITLE
Instagram style modal routing not working with dynamic routes

### DISCUSSION THREAD

**Discussion by mpaus** [vercel/next.js#14789] [Help] [answered]
Hello,

I'm trying to create an instagram style routing.
To be more clear the flow goes like this: 
1. User clicks on an item
2. A modal of the item opens and it adds the items slug to the URL
3. If the page is refreshed (or if you decide to access the item directly through the URL) then a specific page for the selected item opens up

I'm using your "with-route-as-modal-app" example as a reference, to be more specific, this chunk of code is my point of reference:

```
<Link key={index} href={ `/?postId=${id}`} as={`/post/${id}`}>
            <a className={styles.postCard}>{id}</a>
 </Link>

```
Here is an example of how I use it:

```
<Link key={index} href={`/order/[location][category]/?item=${slug}`} as={`/order/los-angeles/phones/${slug}`}>
            <a>{item}</a>
 </Link>

```
The error I get is: `Uncaught (in promise) Error: The provided `as` value (/order/los-angeles/phones/huawei-p30) is incompatible with the `href` value (/order/[location]/[category]).`

If I replace my href value with this: 
`/order/los-angeles/phones/?item=${slug}`
 the said error doesn't happen but it reroutes me to the item page immediately instead of opening the modal.

Can someone please explain to me what I'm doing wrong and what is a possible fix or a workaround?

**Comment by ilackovic** [comment]
I also tried that example repo, but the result was the same. It's either not working properly or it is badly documented. Please explain the OP case

**Comment by Lkriza** [comment]
Interested to know this as well.

**Comment by surferwat** [comment]
Is this helpful? https://github.com/vercel/next.js/issues/8023

**Reply by mpaus** [reply]
Yeah, but I don't like the solution. Basically I would have to change a line in the source code for next router which I don't wanna do.

**Comment by Timer** [comment, marked as answer]
Try this:

```tsx
import { useRouter } from "next/router";

// ...

  const { pathname, query } = useRouter();

  <Link
    key={index}
    shallow
    href={{ pathname, query: { ...query, slug } }}
    as={`/order/los-angeles/phones/${encodeURIComponent(slug)}`}
  >
    <a>{item}</a>
  </Link>
```

I'll mark this as the answer since the provided code in OP shows a common misuse of `<Link>` for shallow routing. Let us know if this doesn't work for your specific use case.

**Reply by mpaus** [reply]
Thanks @Timer!! This solved my issue

**Reply by alaister** [reply]
This still seems to break if I add any sort of query / search string to my `as` url. Are there any work arounds for this?
